function-sidecar/cmd/function-sidecar.go

```diff
@@ -108,31 +111,75 @@ func main() {
 		defer consumer.Close()
 	}
 
-	dispatcher, err := createDispatcher(protocol)
-	if err != nil {
-		panic(err)
-	}
-	switch d := dispatcher.(type) {
-	case io.Closer:
-		log.Print("Requesting close()")
-		defer d.Close()
-	}
-
-	carrier.Run(consumer, producer, dispatcher, output)
-
 	// trap SIGINT and SIGTERM to trigger a shutdown.
 	signals := make(chan os.Signal, 1)
 	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
-	<-signals
-	log.Println(("Shutting Down..."))
+
+LOOP:
+	for {
+
+		select {
+		case <-signals:
+			log.Println("Shutting Down...")
+			break LOOP
+		default:
+		}
+
+		log.Print("Creating dispatcher")
+		dispatcher, err := createDispatcher(protocol)
+		if err != nil {
+			if !backoff() {
+				panic(err)
+			} else {
+				continue LOOP
+			}
+		}
+		switch d := dispatcher.(type) {
+		case io.Closer:
+			log.Print("Requesting close()")
+			defer d.Close()
+		}
+
+		carrier.Run(consumer, producer, dispatcher, output)
+
+		select {
+		case <-signals:
+			log.Println("Shutting Down...")
+			break LOOP
+		case <-dispatcher.Closed():
+			log.Println("End of Stream...")
+			if !backoff() {
```

Comment:
Why back off after end of stream and potentially create another dispatcher? That doesn't seem related to https://github.com/projectriff/riff/issues/323.

Reply:
It's a possibly feeble attempt to address #323. The problem there is that when the invoker either finishes the RPC or just goes away the sidecar sits and does nothing. At least if you try to create another dispatcher, it might be able to have another crack at it. I can't say it's very successful, yet. Hopefully it makes it easier to see how to fix it properly for someone who knows more than I do.